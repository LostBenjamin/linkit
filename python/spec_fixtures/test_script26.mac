# ----------------------------------------------------------------
# Inchworm
# Multiple rainbow wandering inchworms
# ----------------------------------------------------------------

%include symbols
%include test_script26_include

%run-macro inchworm-app

$render-time 100
$width0 `<NUM-LEDS> * 1 / 6`
$width1 <NUM-LEDS>
$width2 `<NUM-LEDS> * 5 / 6`
$time1 50
$time2 200
$seq-type sqs
$fade-type ffd
$num-worms 5

[inchworm-app 10]
  start-application

  # make colors intermingle
  <DRAW-PLUS>
  set-draw-mode

  # schedule the renderer
  (render <render-time>)
  
  # start the worms
  (worms-start)
  
  # start the palette color sequencer
  (hsl-sequencer-start)

# start up all worms
[worms-start]
  # expand wornm start meta-templates
  (((worm-start <num-worms>)))

# expand meta-templates for worms, computing the various values
(((worm <num-worms> `<time1>-(X*<num-worms>)` `<time2>-(X*<num-worms>)` X `X+1` X `X+<num-worms>`)))

# template for starting worms
# X - worm number
[[worm-start X
  (worm-start-X)
]]

# template for all worm processing
# X - worm number
# TIME1 - fast advance time
# TIME2 - slow advance time
# COLOR - palette color index
# STEP - advance step for the slow advancing sequencer
# SEQ1 - fast advancing sequencer number
# SEQ2 - slow advancing sequencer number
[[worm X TIME1 TIME2 COLOR STEP SEQ1 SEQ2
  # start up a worm
  [worm-start-X]
    # allocate a sequencer 1/6 the display width for the small/fast sequence
    {seq-SEQ1},<width0>
    <seq-type>

    # allocate a sequencer for the remaining 5/6 display width to swing the sequence
    {seq-SEQ2},<width2>
    <seq-type>

    # schedule the draw macro
    (worm-draw-X TIME1)
    
    # schedule the advance macro
    (worm-advance-X TIME2)

  # draw the next animation frame
  [worm-draw-X]
    # advance the 1/6 sequencer
    <seq-SEQ1>
    sequence

    # add the two sequencers
    # get and set the offset,window to fill in for this frame
    <seq-SEQ1>,<ADD-SEQ>,<seq-SEQ2>
    sequence-next-window

    # set drawing window
    set-position

    # place the palette color
    COLOR
    add-palette-color

    # set the fade type
    <fade-type>

    # fill the whole space
    flood
    
    # restore the viewport
    reset

  # macro to sequence the 5/6 sequencer at a variable rate
  [worm-advance-X]
    # advance the sequencer
    <seq-SEQ2>,0,STEP
    sequence
]]

