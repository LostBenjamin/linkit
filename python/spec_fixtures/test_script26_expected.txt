10:set:app:1:drw:100,72:sch:31:run:4:run
72:set:flu
4:set:0,360:seq:100,5:sch
5:set:0,0,2:seq:sto:60:rcl:csh
6:set:2,0,1:seq
7:set:4,0,2:seq
8:set:6,0,3:seq
9:set:8,0,4:seq
11:set:10,0,5:seq
12:set:1:seq:1,-5,2:snw:pos:0:pal:ffd:flo:rst
13:set:3:seq:3,-5,4:snw:pos:1:pal:ffd:flo:rst
15:set:5:seq:5,-5,6:snw:pos:2:pal:ffd:flo:rst
17:set:7:seq:7,-5,8:snw:pos:3:pal:ffd:flo:rst
19:set:9:seq:9,-5,10:snw:pos:4:pal:ffd:flo:rst
21:set:1,15:sqs:2,75:sqs:50,12:sch:200,6:sch
23:set:3,15:sqs:4,75:sqs:45,13:sch:195,7:sch
25:set:5,15:sqs:6,75:sqs:40,15:sch:190,8:sch
27:set:7,15:sqs:8,75:sqs:35,17:sch:185,9:sch
29:set:9,15:sqs:10,75:sqs:30,19:sch:180,11:sch
31:set:21:run:23:run:25:run:27:run:29:run

